### DISCUSSION TITLE
Declarative dataclass mapping mixins, relationships, declared_attr, and PEP 681

### DISCUSSION THREAD

**Discussion by yasamoka** [sqlalchemy/sqlalchemy#9408] [Usage Questions] [answered]
First and foremost, thank you to @zzzeek for the wonderful work on SQLAlchemy, especially version 2.0.

Say we have the following mixin classes:

```python
from enum import Enum
from typing import ClassVar, Generic, TypeVar

from sqlalchemy.engine import create_engine
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    MappedAsDataclass,
    declared_attr,
    mapped_column,
    relationship,
)
from sqlalchemy.orm.session import Session
from sqlalchemy.sql.schema import ForeignKey


class Language(Enum):
    EN = "en"
    FR = "fr"


class Base(DeclarativeBase, MappedAsDataclass):
    pass


PersonName = TypeVar("PersonName")


class PersonMixin(Generic[PersonName], MappedAsDataclass):
    __person_name_model_name__: ClassVar[str]

    id: Mapped[int] = mapped_column(init=False, primary_key=True, autoincrement=True)

    @declared_attr
    def names(cls) -> Mapped[list[PersonName]]:
        return relationship(cls.__person_name_model_name__, init=True)


class PersonNameMixin(MappedAsDataclass):
    __person_tablename__: ClassVar[str]

    @declared_attr
    def person_id(cls) -> Mapped[int]:
        return mapped_column(
            ForeignKey(f"{cls.__person_tablename__.lower()}.id"),
            init=False,
            primary_key=True,
        )

    language: Mapped[Language] = mapped_column(init=True, primary_key=True)
    name: Mapped[str] = mapped_column(init=True)
```

`PersonMixin` provides a `Relationship` as a `declared_attr`.

`PersonNameMixin` provides a `ForeignKey` as a `declared_attr`.

Then, we have a set of primary classes:

```python
class Student(PersonMixin["StudentName"], Base):
    __person_name_model_name__ = "StudentName"
    __tablename__ = "student"

    student_id: Mapped[str] = mapped_column(init=True, nullable=False, unique=True)


class StudentName(Base, PersonNameMixin):
    __tablename__ = "student_name"
    __person_tablename__ = "student"
```

`Student` inherits from `PersonMixin`, `StudentName` inherits from `PersonNameMixin`, and now `Student` has a `relationship` to `StudentName`.

Then, we instantiate a `Student`:

```python
student = Student(
    student_id="abc",
    names=[
        StudentName(language=Language.EN, name="John Doe"),
        StudentName(language=Language.FR, name="Jean Lapine"),
    ],
)
```

The code runs successfully. However, pyright complains:

```
(property) def names: (cls: Self@PersonMixin[PersonName@PersonMixin]) -> Mapped[list[PersonName@PersonMixin]]
No parameter named "names" (reportGeneralTypeIssues)
```

Is there a way to get a relationship declared in a mixin with the `@declared_attr` decorator to show up in an inheriting class's `__init__` ([PEP 681](https://peps.python.org/pep-0681/))?

Thank you in advance.

**Comment by zzzeek** [comment, marked as answer]
this is pyright, so you'd need to use a TYPE_CHECKING block that has the relationship() just by itself without the declared attr.

**Reply by yasamoka** [reply]
Wonderful, that works, thank you!!

Out of curiosity, is this a limitation of dataclass transforms (PEP 681) or SQLAlchemy 2.0? Is it because a method is now being used instead of a regular attribute?

**Reply by zzzeek** [reply]
in this specific case it's sort of both.

one way maybe this could be added at some point would be to add "copy from a mixin" support to relationship(), which is a function that Column was given in Declarative early on, and is intrinsic to mapped_column(), but was never done for relationship.   

that would look like:

```py
class PersonMixin(Generic[PersonName], MappedAsDataclass):
    __person_name_model_name__: ClassVar[str]

    id: Mapped[int] = mapped_column(init=False, primary_key=True, autoincrement=True)
    names: Mapped[list[PersonName]] = relationship(lambda cls: cls.__person_name_model_name__, init=True)
```

so each time a mapped class comes up, that relationship would copy itself.   but I can see the awkwardness already that the lambda: is given the "cls", right now relationship accepts lambdas like that but no arguments are passed.

another idea...might be much simpler to implement, make declared_attr have a dataclasses feature itself, like this:

```py
class PersonMixin(Generic[PersonName], MappedAsDataclass):
    __person_name_model_name__: ClassVar[str]

    id: Mapped[int] = mapped_column(init=False, primary_key=True, autoincrement=True)
    names: Mapped[list[PersonName]] = declared_field(init=True)
    
    @names.declared_attr
    @classmethod
    def _make_names(cls) -> Any:
        return relationship(cls.__person_name_model_name__)
```

cc @CaselIT for that second idea

**Reply by yasamoka** [reply]
I see, that makes sense. Thanks again!

**Reply by CaselIT** [reply]
why don't we just make `declared_field` require a callable?

```py
class PersonMixin(Generic[PersonName], MappedAsDataclass):
    __person_name_model_name__: ClassVar[str]

    id: Mapped[int] = mapped_column(init=False, primary_key=True, autoincrement=True)
    # with relationship
    names: Mapped[list[PersonName]] = declared_field(lambda cls: relationship(cls.__person_name_model_name__), init=True)
    # with mapped class or whatever
    has_fk: Mapped[int] = declared_field(lambda cls: mapped_class(ForeignKey(f"{cls.__fk_table__}.id", init=True)

```

maybe this could solve also https://github.com/sqlalchemy/sqlalchemy/issues/9301 and https://github.com/sqlalchemy/sqlalchemy/issues/6120 by adding a flag in `declared_field` to make it invoke after classes definition. something like `declared_field(..., declared_last=True)`

**Reply by zzzeek** [reply]
basically I dont like lambdas very much for things where we expect there's logic going on , like "get the tablename from the class" and such.   this is why I always liked declared_attr as a decorator.

I think Guido is not too keen on lambdas either.  he wanted to remove them in python 3.

**Reply by CaselIT** [reply]
a plain function can be used there too, lambda are just one liners.

what I don't like about the decorator is that you have the type checker screaming at you when you want to use the same name, see hybrid.
here fortunately they scream at you for a reason since everything would break if the function gets the same name as the declared field.

overall I would have a mild preference for a callable here, since I find it clearer, is more compact in most cases and has less possibility of breaking. We could support both without much effort, since a decorator is just a callable anyway

**Reply by CaselIT** [reply]
btw declared_field is more or less the `LazyMapped ` suggested in https://github.com/sqlalchemy/sqlalchemy/issues/9301, just more general

**Comment by surgura** [comment]
Stating this is also a problem with mypy(1.2.0), but as stated by zzzeek it is resolved by using TYPE_CHECKING. Simplified:
```python
if TYPE_CHECKING:
    my_attr: Mapped[T] = mapped_column(nullable=False)
else:
    @declared_attr
    def my_attr(cls) -> Mapped[T]:
        return mapped_column(cls.__type_of_T, nullable=False)
```

Verbose but works fine.

**Reply by surgura** [reply]
Further more, @yasamoka I was able to remove the need for the user to define `__person_name_model_name__` as in your example. See https://stackoverflow.com/questions/76080957/sqlalchemy-2-generic-mixin. It's a bit of a code journey though.

**Reply by zzzeek** [reply]
it's how pep-681 is defined.   it only recognizes callables that are registered with it which have to be in the same configuration as `dataclasses.field()`.

**Reply by rjys365** [reply]
So is there a way that I can quickly define these relationships without adding an if block in each mixin class? Sorry if this sounds unreasonable since I'm new to SQLAlchemy. Thanks a lot.

**Reply by zzzeek** [reply]
there's no update to the answer here, though you could probably roll the solution in https://github.com/sqlalchemy/sqlalchemy/discussions/9408#discussioncomment-5184101 yourself
